Add OpenTracing compatibility section.

carlosalberto opened the pull request:
Fixes #114 

This is a second attempt after #768 was closed ;) It applies all the feedback, leaving the so-called back-propagation of `Baggage` out, given its complexity in the implementation. It can be added later if/as needed:

```java
try (io.opentracing.Scope scope = openTracingTracer.activateSpan(span)) {
  // New baggage item *is properly* updated in Span/Baggage, but the current Context is
  // not updated as of now.
  span.setBaggageItem("foo", "bar");
}
```

cc @pavolloffay @tedsuo @yurishkuro

SergeyKanzhelev commented on specification/compatibility/opentracing.md:
```diff
@@ -0,0 +1,273 @@
+recorded using the OpenTelemetry API. This Shim Layer MUST NOT allow older
+OpenTracing Tracing components to consume upstream OpenTelemetry functionality.
+
+This functionality MUST defined in its own OpenTracing Shim Layer, not in the
```
```suggestion
This functionality MUST be defined in its own OpenTracing Shim Layer, not in the
```

SergeyKanzhelev commented on specification/compatibility/opentracing.md:
```diff
@@ -0,0 +1,273 @@
+`Context` in order to simplify setting the active OpenTelemetry `Span` and
+`Baggage` in a single step.
+
+The `Log` operations MUST be implemented using the OpenTelemetry
```
is there some mapping between well known OT logs and OpenTelemetry semantic convention?

carlosalberto replied:
We only have `error`, which happens at `SetTag`, so I will add it. There were plans to add other members, but that can be done in a follow up (or after GA even).

carlosalberto replied:
Added it as a section at the end (currently we only have `error` - we can add more mappings later on).

SergeyKanzhelev commented on specification/compatibility/opentracing.md:
```diff
@@ -0,0 +1,273 @@
+The `Log` operations MUST be implemented using the OpenTelemetry
+`Span`'s `Add Events` operations.
+
+The `Set Tag` operations MUST be implemented using the OpenTelemetry
```
same qq as for logs - do we need to document any mapping between well-known tags and attributes?

carlosalberto replied:
Same as above.

SergeyKanzhelev commented on specification/compatibility/opentracing.md:
```diff
@@ -0,0 +1,273 @@
+This functionality will be provided as a bridge layer implementing the
+[OpenTracing API](https://github.com/opentracing/specification) using the
+OpenTelemetry API. This layer MUST NOT rely on implementation specific details
+of any SDK.
```
should shim allow both - OT and Otel to work simultaneously for a transition period? I don't think so, but softly. Perhaps it can be said explicitly

carlosalberto replied:
I'm _slightly_ uncomfortable with supporting such scenario, but this is something users are actually asking for, it seems. Yes, worth mentioning it here.

carlosalberto replied:
Added this part ;)

SergeyKanzhelev reviewed (approved):
I'm OK with the document in current form, even though I have a few questions and suggested improvements - like semantical convention mapping

yurishkuro commented on specification/compatibility/opentracing.md:
```diff
@@ -0,0 +1,273 @@
+
+In order to satisfy the OpenTracing `Span` requirements:
+
+- The associated `SpanContext` Shim object will contain a `Context`, which
```
> will contain a `Context`

do you mean OpenTelemetry `Context`?

yurishkuro replied:
Also, why is it a requirement? The shim can extract OTEL Span and Baggage if it needs to.

carlosalberto replied:
So originally added this way as it made it simpler to do activation in a single step, as we already have a `Context`, but I'm having second thoughts about this, as we may leak other values in `Context`. 

Overall I agree that `Span` and `Baggage` can be extracted instead. Sounds like a plan?

carlosalberto replied:
Updated to have `Span` and `Baggage` extracted instead.

yurishkuro commented on specification/compatibility/opentracing.md:
```diff
@@ -0,0 +1,273 @@
+The `Set Tag` operations MUST be implemented using the OpenTelemetry
+`Span`'s `Set Attributes` operations.
+
+### Get SpanContext Shim for OpenTelemetry Span
```
This section does not follow from the document structure. Is it referring to operations of the span shim? Is this some kind of public API of the shim?

carlosalberto replied:
This one and `Set SpanContext Shim for OpenTelemetry Span` are supposed to be Internal operations, i.e. non-public operations:

> This is an INTERNAL helper operation, used by the Shim Layer only. 

... and are defined in order to abstract this step - in turn, this is because each language will decide how to internally track the link between a given OTel Span and a OT SpanContext:

- In Java, we can leverage a `Map` (as shown in the current OT Shim).
- In Python, the OTel Span can have its linked OT SpanContext as `Span._ot_span_context`.

If you think we should make this operations more clear, let me know (maybe add an `Internal` prefix?)

yurishkuro replied:
If it's internal why does it need to be spelled out?  

> how to internally track the link between a given OTel Span and a OT SpanContext

Why is this necessary to track?

yurishkuro replied:
also, if this truly necessary, I would rather have some ability to attach arbitrary runtime-only payloads to OTEL Span than to direct people to use a global map / registry, which are difficult to manage.

carlosalberto replied:
> I would rather have some ability to attach arbitrary runtime-only payloads to OTEL Span than to direct people to use a global map / registry, which are difficult to manage.

For the record, I had initially [proposed this](https://github.com/open-telemetry/opentelemetry-java/pull/596) but this was deemed a bad approach (I agree that handling a global registry can become tricky). We can revisit this as needed for Java and other languages.

> Why is this necessary to track?

Fair enough. Will 'inline' them in the related places.

carlosalberto replied:
Removed them with a very general note on the OTel Span / OT SpanContext relationship.

yurishkuro commented on specification/compatibility/opentracing.md:
```diff
@@ -0,0 +1,273 @@
+- The associated `SpanContext` Shim object will contain a `Context`, which
+  will contain the OpenTelemetry `Span` and `Baggage` values.
+  This `SpanContext` Shim MUST be immutable and MUST be replaced every time
+  baggage is updated through `SetBaggageItem`.
```
SetBaggageItem is never mentioned again in this doc, while it is probably the hardest part of the compatibility because of the inverted relationship between tracer and context propagation in OT vs OTEL.

carlosalberto replied:
Fixed the link, my bad ;(

yurishkuro reviewed (commented):
Do we have a functional prototype of this design?

jmacd commented on specification/compatibility/opentracing.md:
```diff
@@ -0,0 +1,273 @@
+
+More specifically, the intention is to allow OpenTracing instrumentation to be
+recorded using the OpenTelemetry API. This Shim Layer MUST NOT allow older
+OpenTracing Tracing components to consume upstream OpenTelemetry functionality.
```
I couldn't follow this sentence about upstream functionality.

carlosalberto replied:
Amended, let me know - essentially we don't want the OpenTracing Shim to expose in its pubic API any upstream OpenTelemetry functionality, i.e. `Sampler` or `Context`.

SergeyKanzhelev commented:
@carlosalberto I see a couple of comments here. It would be great to make sure @yurishkuro approves this PR as he has way more experience with OT than me.

carlosalberto commented:
> Do we have a functional prototype of this design?

Hey @yurishkuro sorrrry for the super late iteration. So this could inspiration from the [Python take](https://github.com/open-telemetry/opentelemetry-python/blob/master/instrumentation/opentelemetry-instrumentation-opentracing-shim/src/opentelemetry/instrumentation/opentracing_shim/__init__.py#L151) when it came to storing the `Context` right in the OT Shim Layer. Other than that, the OT Java Shim has this approach (storing directly `SpanContext` and `Span` instead).

Let me know. Added a few comments/clarifications. Hopefully we can wrap this up this time ;)

yurishkuro commented on specification/compatibility/opentracing.md:
```diff
@@ -0,0 +1,288 @@
+
+- The associated `SpanContext` Shim object will contain a `Context`, which
+  will contain the OpenTelemetry `Span` and `Baggage` values.
+  This `SpanContext` Shim MUST be immutable and MUST be replaced every time
```
doesn't seem to be the case here: https://github.com/open-telemetry/opentelemetry-python/blob/cd63eb6e52c6f66c9170a2928d2b4dc8918a41de/instrumentation/opentelemetry-instrumentation-opentracing-shim/src/opentelemetry/instrumentation/opentracing_shim/__init__.py#L294

carlosalberto replied:
Correct, this needs to be fixed (similarly how to we need to support the `timestamp` values in the Java one). But yes, we need to definitely make the OT `SpanContext` part immutable.

carlosalberto commented:
Hey @yurishkuro 

Addressed the comments (with the exception of Java's `Span` having a slot to store any Shim related object, and the Python implementation not being up-to-date).

Let me know ;)

yurishkuro commented on specification/compatibility/opentracing.md:
```diff
@@ -0,0 +1,267 @@
+Calls `Set Attribute` on the underlying OpenTelemetry `Span` with the specified
+key/value pair.
+
+The value MUST be [mapped](#semantic-conventions-mapping) to the respective
+OpenTelemetry `Attribute`, or converted to a string if the value type is not supported.
```
I recommend moving that mapping section right here, because it does not apply to anything else and forces the reader to switch context.

Q: are we going to do anything when `sampling.priority` tag is set?

carlosalberto replied:
Not that I'm aware of - we can add it here if you have a clear idea of what should be we mapping it to.

yurishkuro commented on specification/compatibility/opentracing.md:
```diff
@@ -0,0 +1,267 @@
+The value MUST be [mapped](#semantic-conventions-mapping) to the respective
+OpenTelemetry `Attribute`, or converted to a string if the value type is not supported.
+
+If the type of the specified value is not supported, the value MUST be converted
```
> type of the specified value is not supported

Not supported by whom? Would like this to be clearer. In dynamic languages the OTEL SDK would need to deal with any type of value, so the shim does not need to do any additional conversion. It's only when OTEL API in strongly typed languages restricts the types of values for attributes then we may need this to-string coercion (but then the equivalent OT API may not allow `any` value either, e.g. in Java).

carlosalberto replied:
Not supported by OTel, as it is indeed strongly typed. In general, the case was something along "if the value is not a numerical or boolean, simply convert it to string", specially useful for Exception/error objects, for example.

carlosalberto replied:
'Sadly' I realized even the dynamic OTel implementations do check for the type upon setting the `Attribute`, so I'd rather do a conversion at this moment. Filled #1509 as a follow up, in case we update this part (otherwise, we would be dropping values when using the OT Shim, which is *not* something I'd expect).

yurishkuro commented on specification/compatibility/opentracing.md:
```diff
@@ -0,0 +1,267 @@
+the pair set, or else fallback to use the `log` literal string.
+
+The set of values MUST be [mapped](#semantic-conventions-mapping) to the respective
+OpenTelemetry `Attribute`, or converted to a string if the value type is not supported.
```
that section does not contain anything for logs, I would not reference it. If we do need the mapping for logs lets include it here.

yurishkuro commented on specification/compatibility/opentracing.md:
```diff
@@ -0,0 +1,267 @@
+key/value pair set.
+
+The `Add Event`'s `name` parameter MUST be the value with the `event` key in
+the pair set, or else fallback to use the `log` literal string.
```
Where are we with OTEL error/exception reporting? I.e. do we have proper APIs or do we only rely on semantic conventions? 

OT has a specific semantic convention about how errors are supposed to be reported (e.g. `event=error`, etc.), that seems like a conversion should be taking place here to map to OTEL, either via semconv or via proper Errors API.

carlosalberto replied:
We don't rely on semantic conventions at this moment - we rely on a `Status` object at this time.

So, yes - my impression is that we can the `error` events 'go-through' our `AddEvent` operation. Do you have something else in mind? My impression is that the **actual** way to properly set an error is through `span.setTag("error", true)`, which we handle properly at this moment.

carlosalberto replied:
I suggest we do this in a follow-up, as in theory we should be able to 'translate' somehow a specific `Log(Exception)` call to a `RecordException(Exception)` call, but the details need to be analyzed.

yurishkuro commented on specification/compatibility/opentracing.md:
```diff
@@ -0,0 +1,267 @@
+If an explicit timestamp is specified, a conversion MUST be done to match the
+OpenTracing and OpenTelemetry units.
+
+### SpanContext Shim
```
move to a section at the same level as Span Shim

yurishkuro commented on specification/compatibility/opentracing.md:
```diff
@@ -0,0 +1,267 @@
+
+#### Get Baggage Items
+
+Returns a Shim object exposing the associated OpenTelemetry `Baggage` values.
```
"Shim" is a bit weird here, in all OT APIs this either returns an iterator or a well-defined collection.

```suggestion
Returns a dictionary, collection, or iterator (depending on the requirements of the OpenTracing API for a specific language)
backed by the associated OpenTelemetry `Baggage` values.
```

carlosalberto replied:
In OT Java we do need a Shim, but that may be a specific case. Adjusting now, thanks.

yurishkuro commented on specification/compatibility/opentracing.md:
```diff
@@ -0,0 +1,267 @@
+
+Gets the [associated](#opentelemetry-span-and-spancontext-relationship)
+`SpanContext` Shim for the specified `Span` and puts its OpenTelemetry
+`SpanContext` and `Bagagge` in a new `Context`, which is then set as the
```
iirc the OTEL Context stores OTEL Span, not Span Context.

carlosalberto replied:
Yes, definitely, thanks ;) I totally oversaw this.

yurishkuro commented on specification/compatibility/opentracing.md:
```diff
@@ -0,0 +1,267 @@
+
+### Get the active Span
+
+Gets the active OpenTelemetry `Span` and returns a `Span` Shim wrapping it.
```
This is a tricky one. Are we saying a new Span Shim must be constructed at this point, or that the Span Shim should be stored in the Context? I think for performance reasons we want the latter.

carlosalberto replied:
Try to put a `SpanShim` into a `Context` is an interesting optimization. Still, for cases where activation didn't happen _from_ the OT side, we would still need to create it:

```java
static void myMethod1() {
  // OTel api
  try (Scope scope = span.makeCurrent()) {
    myMethod2();
  }
}

static void myMethod2() {
  // No way OTel can know about the shim.
  io.opentracing.Span span = GlobalTracer.get().activeSpan();
}
```

As it sounds like an easy optimization (trying to put/get the `SpanShim` in the `Context` by default), I can definitely add it, if that sounds ok. Please confirm if this sounds fine with you.

carlosalberto commented on specification/compatibility/opentracing.md:
```diff
@@ -0,0 +1,267 @@
+### OpenTelemetry Span and SpanContext relationship
+
+A given OpenTelemetry `Span` MUST be associated with only one
+`SpanContext` Shim at all times for all execution units, in order
```
I think we use it in other places - essentially we wanted to get far from using `threads` as there are similar components such as coroutines in Python.

carlosalberto replied:
Filled #1511

jsuereth commented:
Hey @carlosalberto what's the status on this PR?  I'm working on contributing something similar for OpenCensus, and wanted to make sure this PR was close to complete, as I was planning to line up the strategies and verbage quite similarly.

Thanks!

carlosalberto commented:
@jsuereth Hey! The plan is to iterate on it this week. The 'only' remaining things to complete are the currently standing comments ;)

SergeyKanzhelev commented:
Removed stale

carlosalberto commented on specification/compatibility/opentracing.md:
```diff
@@ -0,0 +1,267 @@
+- `TEXT_MAP` and `HTTP_HEADERS` formats MUST use the registered OpenTelemetry
+  `HTTPTextPropagator`, if any.
+
+Errors MUST NOT be raised if the specified `Format` is `BINARY`.
```
>  if a propagator is not configured for a certain OT format, then a fallback behavior can be prescribed.

Makes sense. 

>  e.g. in Go an error is explicitly returned:

Good call. Will adjust to allow this.

carlosalberto commented on specification/compatibility/opentracing.md:
```diff
@@ -0,0 +1,267 @@
+Inject the underlying OpenTelemetry `Span` and `Bagagge` using the registered
+OpenTelemetry `Propagator`s:
+
+- `TEXT_MAP` and `HTTP_HEADERS` formats MUST use the registered OpenTelemetry
```
Is [this](https://github.com/jaegertracing/jaeger-client-java/blob/master/jaeger-core/src/main/java/io/jaegertracing/internal/propagation/TextMapCodec.java#L203) the encoding you mention? If yes, I created a prototype that allows specifying encoding/decoding: https://github.com/open-telemetry/opentelemetry-java/compare/master...carlosalberto:ot_shim_encoded?expand=1

But if we ever support this encoding in our OTel Jaeger Propagators, then we would be repeating the conversion, which sounds bad. In that case, we could go with the approach you mentioned in another comment:

```go
otTracer := createTracerShim(otelTracer, propagators = {
    ot.TEXT_MAP: otelsdk.getJaegerTextMapPropagator(),
    ot.HTTP_HEADERS: otel.getHTTPTextPropagator(),
    ot.BINARY: null,
})
```

If not specified, we simply use OTel TextMap for both TextMap/HttpHeaders. Any opinion on this one? I'm fine with either of them.

yurishkuro replied:
Not quite following why we'd be repeating conversion - OT shim won't do anything by itself, just delegate to OTEL propagator. I am just saying that there needs to be a way to use different propagators for TEXT_MAP vs. HTTP_HEADERS formats, contrary to what this L76 says.

Your branch is kind of going in this direction, but not quite, because this:

```java
public static io.opentracing.Tracer createTracerShim(
      OpenTelemetry openTelemetry, boolean urlEncoding) {
```

_fixes_ the urlEncoding for a given OT tracer. But the whole point is that whether to apply urlEncoding or not is not a fixed property of the tracer, but the side effect of which format the user is trying use to to encode the context. A single tracer can be used in a service that has both HTTP server _and_ publishes to Kafka, and that service would use different OT formats for those two trace points.

Passing a map of propagators instead of a boolean flag when creating a shim should resolve this generically.

carlosalberto replied:
Hey @yurishkuro I'm fine allowing the user to pass a set of propagators at OT Shim creation time - then the behavior should be IMHO:

a) If user specified any Propagator, have the Shim use them for **all** propagation, ignoring the global Propagators.
b) If the user didn't specify any Propagator, simply fallback to use the global Propagators.

Is that a good compromise for you? Or think that for a) we should _also_ override the global Propagators?

yurishkuro replied:
I don't think creating a shim should affect global propagators.

I would combine a) and b) into this:
  * user can pass specific propagators for each of the OpenTracing formats
  * if a propagator is not specified for one of the OT formats, the shim will fall back to default OTEL propagator as follows:
    * OT.TEXT_MAP -> OTEL.HTTP  
    * OT.HTTP_HEADERS -> OTEL.HTTP
    * OT.BINARY -> not supported at this time

There is also a question on the typing: in OTEL the text and binary propagators are likely to have different signatures.

carlosalberto replied:
Thanks, will update the appropriately, along with a small prototype for this (just as done before).

carlosalberto commented on specification/compatibility/opentracing.md:
```diff
@@ -0,0 +1,267 @@
+
+The OpenTelemetry project aims to provide backwards compatibility with the
+[OpenTracing](https://opentracing.io) project in order to ease migration of
+instrumented codebases.
```
Actually I mention it below: https://github.com/open-telemetry/opentelemetry-specification/pull/1101/files#diff-efb9735083a18d449551326756460206cf789d785e16e4a7301c9eb3e0205f36R46

carlosalberto commented on specification/compatibility/opentracing.md:
```diff
@@ -0,0 +1,267 @@
+## Span Shim
+
+The `Span` operations MUST be implemented using underlying OpenTelemetry `Span`
+and `Baggage` values through a `SpanContext` Shim.
```
I recurrently mention a **SpanContext Shim** as this is the 'core' object for the `SpanContext`/`Baggage` composite object. I can leave this out but I think there's value in having, at the document level, as a 'first class' citizen, in order to clarify the Shim expectations.

carlosalberto commented on specification/compatibility/opentracing.md:
```diff
@@ -0,0 +1,267 @@
+The `Span` operations MUST be implemented using underlying OpenTelemetry `Span`
+and `Baggage` values through a `SpanContext` Shim.
+
+In order to satisfy the OpenTracing `Span` requirements:
```
All right, this can be moved to its own section.

carlosalberto commented on specification/compatibility/opentracing.md:
```diff
@@ -0,0 +1,267 @@
+to keep any linked `Baggage` consistent across all execution
+units at all times.
+
+An example of this scenario is having an OpenTracing `Span` have its
+[Set Baggage Item](#set-baggage-item) operation called from two different
+execution units (e.g. threads), and afterwards have its [Context](#get-context)
+fetched in order to [iterate over its baggage values](#get-baggage-items).
+
+Getting and setting the currently associated `SpanContext` for a specified
+OpenTelemetry `Span` MUST be safe to call from different execution units.
+
+Managing this relationship is an implementation detail. It can be implemented,
+for example, with the help of a global synchronized dictionary, or with an
+additional attribute in OpenTelemetry `Span` objects for dynamic languages.
```
In Java and Python `Baggage` exists in its own object, which basically means we can leave `Context` entirely out (at least for those languages). 

> Specifically, in OT the span.setBaggage operation is change-in-place and affects what baggage can be seen by different threads

This is indeed the reason I'm mentioning here the details. However, I'm *also* trying to keep the invariant of an actual `SpanContext` being entirely replaced:

```java
// OT perspective
SpanContext ctx1 = span.getContext();
span.setBaggageItem("hey", "you");
assertNotEquals(ctx1, span.getContext());
```

If you think this invariant is not worth keeping (and it's enough to simply keep the *actual values* consistent), I can try to play around and simplify things.

carlosalberto commented:
@yurishkuro Sorry for the late feedback (had the Christmas break in between). I will iterate on the changes once you leave your own feedback for 1-2 of the remaining tricky issues. Thanks!

bogdandrutu commented on specification/compatibility/opentracing.md:
```diff
@@ -0,0 +1,268 @@
+# OpenTracing Compatibility
+
+<details>
+<summary>Table of Contents</summary>
+
```
Please add status

reyang commented on specification/compatibility/opentracing.md:
```diff
@@ -0,0 +1,268 @@
+
+## ScopeManager Shim
+
+For OpenTracing languages implementing the `ScopeManager` interface,  its operations
```
nit:

```suggestion
For OpenTracing languages implementing the `ScopeManager` interface, its operations
```

reyang commented on specification/compatibility/opentracing.md:
```diff
@@ -0,0 +1,268 @@
+- A set of key/value pairs, where keys must be strings, and the values may have
+  any type.
+
+Calls `Add Event` on the underlying OpenTelemetry `Span` with the specified
```
minor: consider `Add Events` (plural form) to align with the [spec](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#add-events)?

reyang commented on specification/compatibility/opentracing.md:
```diff
@@ -0,0 +1,268 @@
+The `Add Event`'s `name` parameter MUST be the value with the `event` key in
+the pair set, or else fallback to use the `log` literal string.
+
+The set of values MUST be [mapped](#semantic-conventions-mapping) to the respective
```
Here we didn't specify what to do if the value passed in via the Shim is null (assuming OpenTracing will discard the key/value pair if value is null?).

carlosalberto replied:
Oh, good call. I think we should filter them out, to stay defensive. Will fill a follow up for that so the entire doc is reviewed for that part.

reyang reviewed (approved):
LGTM with some minor suggestions.

carlosalberto commented:
Hey @yurishkuro 

I finally managed to update the document as suggested in the previous review. A few things to summarize the state:

- Overall re-organized the document following your recommendations.
- Filled an issue for the "execution unit" term.
- Allowed the Tracer Shim to get explicit, separate propagators for OT `TextMap` and `HttpHeaders` formats. Binary is not mentioned at all for now, as we don't quite define it anywhere.
- Added code snippets for non-obvious operations - although went with Java, as I also wanted to describe `ScopeManager`, which doesn't exist in Go ;)
- Will work on the `Log(Exception e)` -> `RecordEvent(Exception e)` translation in a follow up, in order to facilitate merging this PR.

The final important item to be discussed is your comment:

> Specifically, in OT the span.setBaggage operation is change-in-place and affects what baggage can be seen by different threads (and, therefore, has an inherent race condition, due to OT design). In contrast, OTEL stores baggage in the Context and removes this race condition by design, but it is not clear to be how the shim deals with it since the Context is never mentioned. If we carry the Baggage object directly in the SpanContext shim, then it can reproduce the same OT behavior.

To which I had answered:

> In Java and C# Baggage exists in its own object, which basically means we can leave Context entirely out (at least for those languages)...  I'm also trying to keep the invariant of an actual SpanContext being entirely replaced (by calling Span.setBaggageItem()) 

Let me know ;)

carlosalberto commented:
Merging as we have enough approvals, and have filled the following issues as follow ups:

#1511 Execution Unit clarification
#1552 Error logging translation
#1553 Bagage storage location

@yurishkuro please feel free to open any other follow up, I will be happy to address it.

I think now we need to go out and validate with Go/Python to finish validating (or correcting) the current approach.
